remotes/docker/fetcher.go

```diff
@@ -151,12 +152,18 @@ func (r dockerFetcher) Fetch(ctx context.Context, desc ocispec.Descriptor) (io.R
 	})
 }
 
-func (r dockerFetcher) createGetReq(ctx context.Context, host RegistryHost, ps ...string) (*request, int64, error) {
+func (r dockerFetcher) createGetReq(ctx context.Context, host RegistryHost, mediatype string, ps ...string) (*request, int64, error) {
 	headReq := r.request(host, http.MethodHead, ps...)
 	if err := headReq.addNamespace(r.refspec.Hostname()); err != nil {
 		return nil, 0, err
 	}
 
+	if mediatype == "" {
+		headReq.header.Set("Accept", "*/*")
+	} else {
+		headReq.header.Set("Accept", strings.Join([]string{mediatype, `*/*`}, ", "))
+	}
```

Comment:
Why are we manually constructing a comma-separated string here? http header allows multiple headers for the same type;

```
headReq.header.Add("Accept", "*/*")
```

Were we previously setting `*/*` as header? it none was set?


That said, I have doubts about this change; if this is for fetching blobs, then the blob.. is a blob.. there is no content-negotiation specified as part of the endpoint? Aren't an opaque blob of data? (which is why they're uploaded with `application/octet-stream` (https://www.rfc-editor.org/rfc/rfc2046.html#section-4.5.1)?

https://github.com/opencontainers/distribution-spec/blob/v1.1.0-rc3/spec.md#pulling-blobs

Reply:
thank you for reviewing sir! 
· for the header.Set, I followed format in dockerFetcher.open. But I think you right, using Add work as well, will raise a pr to change it for community to review
· previously none was set, and some registry complains 404 #8743 
· I think you are absolutely right that invoker of FetchByDigest should be trying to obtain an opaque blob of data (might be manifest or blob) referenced by the specified digest (i think thats why Suda originally classified returned digest as `application/octet-stream`). But caller needs an option to specify header for registry, otherwise registry might potentially complain 404 for manifest and there's no way for client to make it work (previously fetch-blob work for manifest of nginx:latest but doesn't work for ubuntu:latest) I think the reason might be that in [spec](https://github.com/opencontainers/distribution-spec/blob/v1.1.0-rc3/spec.md#pulling-manifests) it is said that client `SHOULD` instead of `MUST`..
again, thx for reviewing :)

Reply:
Hm... right, so I guess there's two things here;

> previously none was set, and some registry complains 404 #8743

That looks like a bug on Docker Hub's side. Looking at [RFC2616, Section 14.1](https://www.ietf.org/rfc/rfc2616.html#section-14.1):

    If no Accept header field is present, then it is assumed that the
    client accepts all media types. If an Accept header field is present,
    and if the server cannot send a response which is acceptable
    according to the combined Accept field value, then the server SHOULD
    send a 406 (not acceptable) response.

So it should

- consider a missing header to be the same as `*/*`
- IF an `Accept` header is set, and it can't provide content, then it should return a `406`, not a `404`

HOWEVER, I would expect that to only apply to the `manifests` endpoint (`GET /v2/<name>/manifests/<reference>`), and not for blobs (`GET /v2/<name>/blobs/<digest>`). For _manifests_, the registry SHOULD return the original content-type, but for blobs, the content type should (must?) always be considered `application/octet-stream`


>  I think you are absolutely right that invoker of FetchByDigest should be trying to obtain an opaque blob of data (might be manifest or blob)

For the blob endpoint (`GET /v2/<name>/blobs/<digest>`), it should not matter what's pulled; regardless  if it was a manifest, or a layer, or a GIF, it's always a "blob", and always `application/octet-stream`

However, this function (`createGetReq`) is used for both;

- https://github.com/containerd/containerd/blob/40f26543bdc27cbe8b058ac082e91c5832bb1c41/remotes/docker/fetcher.go#L212
- https://github.com/containerd/containerd/blob/40f26543bdc27cbe8b058ac082e91c5832bb1c41/remotes/docker/fetcher.go#L225

And currently, it appenda `*/*` to the header, which would be problematic if the client makes a request for manifests, but does _NOT_ support all manifest-types.

I think this function should not apply headers on its own.

Reply:
>> previously none was set, and some registry complains 404 https://github.com/containerd/containerd/issues/8743
>
> That looks like a bug on Docker Hub's side. Looking at [RFC2616, Section 14.1](https://www.ietf.org/rfc/rfc2616.html#section-14.1):


I checked with a colleague who works on Docker Hub registry, and they found that this issue is unlikely related to the header. In the steps in https://github.com/containerd/containerd/issues/8743, you're trying to fetch a blob from the _ubuntu_ repository, but that blob is _NOT_ referenced by any of the Ubuntu images. In that case, the registry will always respond with a 404 as per specification.

Can you try if the issue also reproduces when fetching a blob that's actually referenced from any of the Ubuntu images?

Reply:
thx for checking, pls check the screenshot in #8743
the layer definitely exists since after including the accept header, registry does not complain 404 anymore.